File: apps/movil/lib/pantallas/soporte/vista_de_tutoriales.dart
import '../../dominio/tutoriales/avance.dart';
import '../../dominio/tutoriales/modelo.dart';

/// Los filtros del centro de ayuda.
enum FiltroDeEstado { todos, pendiente, enProgreso, completado }

/// Una tarjeta ya resuelta: el tutorial más lo que le pasó a esta persona.
class TutorialEnLista {
  const TutorialEnLista({
    required this.tutorial,
    required this.estado,
    required this.fraccion,
    required this.continuable,
    required this.progreso,
    required this.requisitosPendientes,
  });

  final TutorialDefinicion tutorial;
  final EstadoDeProgreso estado;
  final double fraccion;
  final bool continuable;
  final ProgresoDeTutorial? progreso;
  final List<TutorialDefinicion> requisitosPendientes;
}

/// **Lo que el centro de ayuda muestra, calculado sin pintar nada.**
///
/// Función pura sobre tres entradas —catálogo, avance y filtro— así que el orden de la
/// lista y el texto de cada botón se prueban con listas literales, sin levantar la app.
List<TutorialEnLista> componerLista(
  List<TutorialDefinicion> tutoriales,
  Map<String, ProgresoDeTutorial> progresos, {
  String texto = '',
  FiltroDeEstado filtro = FiltroDeEstado.todos,
  String? categoria,
}) {
  final porId = {for (final t in tutoriales) t.id: t};
  final lista = tutoriales
      .map(
        (t) => TutorialEnLista(
          tutorial: t,
          estado: estadoDe(progresos[t.id], t),
          fraccion: fraccionDe(progresos[t.id], t),
          continuable: sePuedeContinuar(progresos[t.id], t),
          progreso: progresos[t.id],
          requisitosPendientes: requisitosPendientes(t, porId, progresos),
        ),
      )
      .where((f) => _coincide(f, texto, filtro, categoria))
      .toList();
  lista.sort((a, b) => _peso(a).compareTo(_peso(b)));
  return lista;
}

/// El que conviene hacer ahora: el primero sin hacer y sin requisitos pendientes.
/// `null` cuando no queda nada, que también es una respuesta.
TutorialEnLista? recomendado(List<TutorialEnLista> lista) {
  for (final f in lista) {
    if (f.estado != EstadoDeProgreso.completado &&
        f.requisitosPendientes.isEmpty) {
      return f;
    }
  }
  return null;
}

bool _coincide(
  TutorialEnLista fila,
  String texto,
  FiltroDeEstado filtro,
  String? categoria,
) {
  if (categoria != null && fila.tutorial.categoria != categoria) return false;
  if (!_coincideEstado(fila, filtro)) return false;
  final busqueda = texto.trim().toLowerCase();
  if (busqueda.isEmpty) return true;
  final heno =
      '${fila.tutorial.titulo} ${fila.tutorial.descripcion} ${fila.tutorial.categoria}'
          .toLowerCase();
  return busqueda.split(RegExp(r'\s+')).every(heno.contains);
}

bool _coincideEstado(TutorialEnLista fila, FiltroDeEstado filtro) =>
    switch (filtro) {
      FiltroDeEstado.todos => true,
      // Omitido cuenta como pendiente: se dejó a medias y sigue sin hacerse.
      FiltroDeEstado.pendiente =>
        fila.estado == EstadoDeProgreso.pendiente ||
            fila.estado == EstadoDeProgreso.omitido,
      FiltroDeEstado.enProgreso => fila.estado == EstadoDeProgreso.enProgreso,
      FiltroDeEstado.completado => fila.estado == EstadoDeProgreso.completado,
    };

/// Primero lo importante, después lo que está a medias, después lo pendiente, y al
/// final lo ya hecho: el orden en el que a uno le sirve encontrarlo.
int _peso(TutorialEnLista fila) {
  if (fila.estado == EstadoDeProgreso.completado) return 4;
  if (fila.tutorial.obligatorio) return 0;
  if (fila.estado == EstadoDeProgreso.enProgreso) return 1;
  return fila.estado == EstadoDeProgreso.omitido ? 3 : 2;
}

File: packages/tutoriales/src/centro/vista-de-tutoriales.ts
import { estadoDe, fraccionDe, requisitosPendientes, sePuedeContinuar } from '../avance'
import type { EstadoDeProgreso, ProgresoDeTutorial, TutorialDefinicion } from '../tipos'

/** Los filtros del centro de ayuda, tal como viven en la dirección. */
export type FiltroDeEstado = 'todos' | 'pendiente' | 'en-progreso' | 'completado' | 'obligatorio'

export interface Filtro {
  readonly texto: string
  readonly estado: FiltroDeEstado
  readonly categoria: string | null
}

/** Una tarjeta, ya resuelta: el tutorial más lo que le pasó a esta persona. */
export interface TutorialEnLista {
  readonly tutorial: TutorialDefinicion
  readonly estado: EstadoDeProgreso
  readonly fraccion: number
  readonly continuable: boolean
  readonly progreso: ProgresoDeTutorial | undefined
  readonly requisitosPendientes: readonly TutorialDefinicion[]
}

/**
 * **Lo que el centro de ayuda muestra, calculado sin pintar nada.**
 *
 * Es una función pura sobre tres entradas —catálogo, avance y filtro— así que el orden
 * de la lista y el texto de cada botón se prueban con arreglos literales, sin montar un
 * componente ni tocar el DOM.
 */
export function componerLista(
  tutoriales: readonly TutorialDefinicion[],
  progresos: ReadonlyMap<string, ProgresoDeTutorial>,
  filtro: Filtro,
): readonly TutorialEnLista[] {
  const porId = new Map(tutoriales.map((t) => [t.id, t]))
  return tutoriales
    .map<TutorialEnLista>((tutorial) => ({
      tutorial,
      estado: estadoDe(progresos.get(tutorial.id), tutorial),
      fraccion: fraccionDe(progresos.get(tutorial.id), tutorial),
      continuable: sePuedeContinuar(progresos.get(tutorial.id), tutorial),
      progreso: progresos.get(tutorial.id),
      requisitosPendientes: requisitosPendientes(tutorial, porId, progresos),
    }))
    .filter((fila) => coincide(fila, filtro))
    .sort(porPrioridad)
}

/**
 * El que conviene hacer ahora: el primero obligatorio o a medias que no tenga
 * requisitos pendientes. `undefined` cuando no queda nada por hacer, que también es una
 * respuesta.
 */
export function recomendado(lista: readonly TutorialEnLista[]): TutorialEnLista | undefined {
  return lista.find((f) => f.estado !== 'completado' && f.requisitosPendientes.length === 0)
}

function coincide(fila: TutorialEnLista, filtro: Filtro): boolean {
  if (filtro.categoria !== null && fila.tutorial.categoria !== filtro.categoria) return false
  if (!coincideEstado(fila, filtro.estado)) return false
  const texto = (filtro.texto ?? '').trim().toLowerCase()
  if (texto.length === 0) return true
  const heno = `${fila.tutorial.titulo} ${fila.tutorial.descripcion} ${fila.tutorial.categoria}`.toLowerCase()
  return texto.split(/\s+/).every((palabra) => heno.includes(palabra))
}

function coincideEstado(fila: TutorialEnLista, estado: FiltroDeEstado): boolean {
  switch (estado) {
    case 'todos':
      return true
    case 'obligatorio':
      return fila.tutorial.obligatorio === true
    case 'pendiente':
      // Omitido cuenta como pendiente: se dejó a medias y sigue sin hacerse.
      return fila.estado === 'pendiente' || fila.estado === 'omitido'
    default:
      return fila.estado === estado
  }
}

/**
 * Primero lo obligatorio, después lo que está a medias, después lo pendiente, y al
 * final lo ya hecho: el orden en el que a uno le sirve encontrarlo.
 */
function porPrioridad(a: TutorialEnLista, b: TutorialEnLista): number {
  return peso(a) - peso(b)
}

function peso(fila: TutorialEnLista): number {
  if (fila.estado === 'completado') return 4
  if (fila.tutorial.obligatorio === true) return 0
  if (fila.estado === 'en-progreso') return 1
  return fila.estado === 'omitido' ? 3 : 2
}
